server/src/main/java/org/elasticsearch/ingest/IngestService.java

```
@@ -569,16 +574,37 @@ public void validatePipelineRequest(ProjectId projectId, PutPipelineRequest requ
         validatePipeline(ingestInfos, projectId, request.getId(), config);
     }
 
-    public static boolean isNoOpPipelineUpdate(ProjectMetadata metadata, PutPipelineRequest request) {
+    public static Map<String, Object> readPipelineConfig(PutPipelineRequest request) {
+        return XContentHelper.convertToMap(request.getSource(), false, request.getXContentType()).v2();
+    }
+
+    public static void validateNoSystemPropertiesInPipelineConfig(final Map<String, Object> pipelineConfig) {
+        if (pipelineConfig.containsKey(Pipeline.CREATED_DATE_KEY)) {
+            throw new ElasticsearchParseException("Provided a pipeline property which is managed by the system: created_date.");
+        } else if (pipelineConfig.containsKey(Pipeline.MODIFIED_DATE_KEY)) {
+            throw new ElasticsearchParseException("Provided a pipeline property which is managed by the system: modified_date.");
+        }
+    }
+
+    public static boolean isNoOpPipelineUpdate(
+        ProjectMetadata metadata,
+        PutPipelineRequest request,
+        Supplier<Map<String, Object>> newPipelineConfigSupplier
```

With this change of signature, it's not at all clear to me what this method does any more. If you're going to make this change, I think it now needs javadoc.

However, I think there's probably a more elegant way to do it, unless I'm misunderstanding. As far as I can see, this is here so that you can do the `convertToMap` outside of this method, so that you can do validation there without repeating that work, is that right? Couldn't you avoid this by doing it as part of the existing `validatePipelineRequest` method? (I think the behaviour will actually be identical — you're moving it after the `isNoOpPipelineUpdate` call, but that would always return false if the request set either property because you're blanking them out from the map you're comparing to, right?)

---

Incidentally, we seem to be happy enough do the `convertToMap` up to three times during the existing pipeline, so it wouldn't actually be the end of the world to do it a fourth... 🤷

---

> so that you can do validation there without repeating that work, is that right?

yeah my thinking was
1. we already deserialize multiple times, two diametric choices are I either add another deserialization, or refactor a bit more to only ever deserialize once; went with half-measure to avoid another deserialization to at least not make the world worse
2. fail early if user provides non-valid properties
3. i already might need to deserialize within `isNoOpPipelineUpdate`, and I need to be aware of system properties there (if i don't assert incoming request doesn't have system properties before function, i'll also need to remove them from the incoming request to check for NOP), so made sense to put it close to it, and `isNoOpPipelineUpdate` exists before going into the listener where `validatePipelineRequest` is

thinking i'll run through this with you in our 1:1